cache/cache_reservation_manager.h

```diff
@@ -26,7 +26,8 @@
 namespace ROCKSDB_NAMESPACE {
 // CacheReservationManager is an interface for reserving cache space for the
 // memory used
-class CacheReservationManager {
+class CacheReservationManager :
+      public std::enable_shared_from_this<CacheReservationManager>{
```

Comment:
[Note] we need to store pointer to CacheReservationManager in CacheReservationHandle so we can't really do "std::unique_ptr<CacheReservationManager>" in call-sites, enforced by compilation error from enable_shared_from_this if do so (hence extra changes included in this PR) 

Benchmark will be included for those changes for concerns on the slow-down from shared_ptr. 

A relevant discussion https://github.com/facebook/rocksdb/pull/9073/files#r748641039 - I can't really think of a way to improve this so I'll leave it as it is right now but welcome any ideas/thoughts!